tests/evm_backends/base_env.py

```diff
@@ -0,0 +1,228 @@
+        *args,
+        input_bundle=None,
+        initcode_prefix=ERC5202_PREFIX,
+    ):
+        """Deploy a contract with a blueprint pattern."""
+        abi, bytecode = _compile(source_code, output_formats, compiler_settings, input_bundle)
+        bytecode = initcode_prefix + bytecode
+        bytecode_len = len(bytecode)
+        bytecode_len_hex = hex(bytecode_len)[2:].rjust(4, "0")
+        # prepend a quick deploy preamble
+        deploy_preamble = bytes.fromhex("61" + bytecode_len_hex + "3d81600a3d39f3")
+        deploy_bytecode = deploy_preamble + bytecode
+
+        deployer_abi: list[dict] = []  # just a constructor
+        value = 0
+        deployer = self.deploy(deployer_abi, deploy_bytecode, value, *args)
+
+        def factory(address):
+            return ABIContractFactory.from_abi_dict(abi).at(self, address)
+
+        return deployer, factory
+
+    def get_logs(self, contract: ABIContract, event_name: str = None, raw=False):
+        logs = [log for log in self.last_result.logs if contract.address == log.address]
+        if raw:
+            return [log.data for log in logs]
+
+        parsed_logs = [contract.parse_log(log) for log in logs]
+        if event_name:
+            return [log for log in parsed_logs if log.event == event_name]
+
+        return parsed_logs
+
+    @contextmanager
+    def anchor(self):
+        raise NotImplementedError  # must be implemented by subclasses
+
+    def get_balance(self, address: str) -> int:
+        raise NotImplementedError  # must be implemented by subclasses
+
+    def set_balance(self, address: str, value: int):
+        raise NotImplementedError  # must be implemented by subclasses
+
+    @property
+    def accounts(self) -> list[str]:
+        raise NotImplementedError  # must be implemented by subclasses
+
+    @property
+    def block_number(self) -> int:
+        raise NotImplementedError  # must be implemented by subclasses
+
+    @property
+    def timestamp(self) -> int | None:
+        raise NotImplementedError  # must be implemented by subclasses
+
+    @property
+    def last_result(self) -> ExecutionResult:
+        raise NotImplementedError  # must be implemented by subclasses
+
+    def execute_code(
+        self,
+        to: str,
+        sender: str | None = None,
+        data: bytes | str = b"",
+        value: int = 0,
+        gas: int | None = None,
+        gas_price: int = 0,
+        is_modifying: bool = True,
+        blob_hashes: Optional[list[bytes]] = None,  # for blobbasefee >= Cancun
+    ) -> bytes:
+        raise NotImplementedError  # must be implemented by subclasses
+
+    def clear_transient_storage(self) -> None:
+        raise NotImplementedError  # must be implemented by subclasses
+
+    def get_code(self, address: str) -> bytes:
+        raise NotImplementedError  # must be implemented by subclasses
+
+    def time_travel(self, num_blocks=1) -> None:
+        raise NotImplementedError  # must be implemented by subclasses
```

Comment:
i don't think the primitive on the env should be `time_travel` -- i think this has a clear implementation:
```suggestion
    def time_travel(self, num_blocks=1) -> None:
        timedelta = num_blocks * 12
        self.block_timestamp += timedelta  # could use getter/setter instead
        self.block_number += num_blocks  # ditto
```

Reply:
in other words, get/set block number and timestamp should be abstract, not `time_travel`